From: Mauro Carvalho Chehab <mchehab@redhat.com>
Date: Tue, 23 Nov 2010 17:33:47 -0500
Subject: [edac] i7core_edac: remove PCI devices from devices list
Message-id: <20101123153347.4cc0cb53@pedra>
Patchwork-id: 29563
O-Subject: [PATCH RHEL5 07/29] BZ#:651869 i7core_edac: explicitly remove PCI
	devices from the devices list
Bugzilla: 651869

Changeset: 39300e7143f8ef81b07cee3d8b86880bc4311ea0

Signed-off-by: Mauro Carvalho Chehab <mchehab@redhat.com>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/edac/i7core_edac.c b/drivers/edac/i7core_edac.c
index f1c53db..c8dbeba 100644
--- a/drivers/edac/i7core_edac.c
+++ b/drivers/edac/i7core_edac.c
@@ -1241,16 +1241,17 @@ static void i7core_put_devices(struct i7core_dev *i7core_dev)
 		pci_dev_put(pdev);
 	}
 	kfree(i7core_dev->pdev);
-	list_del(&i7core_dev->list);
-	kfree(i7core_dev);
 }
 
 static void i7core_put_all_devices(void)
 {
 	struct i7core_dev *i7core_dev, *tmp;
 
-	list_for_each_entry_safe(i7core_dev, tmp, &i7core_edac_list, list)
+	list_for_each_entry_safe(i7core_dev, tmp, &i7core_edac_list, list) {
 		i7core_put_devices(i7core_dev);
+		list_del(&i7core_dev->list);
+		kfree(i7core_dev);
+	}
 }
 
 static void __init i7core_xeon_pci_fixup(const struct pci_id_table *table)
@@ -1439,7 +1440,6 @@ static int i7core_get_devices(const struct pci_id_table *table)
 	}
 
 	return 0;
-	return 0;
 }
 
 static int mci_bind_devs(struct mem_ctl_info *mci,
@@ -2079,6 +2079,8 @@ static void __devexit i7core_remove(struct pci_dev *pdev)
 			/* Release PCI resources */
 			i7core_put_devices(i7core_dev);
 		}
+		list_del(&i7core_dev->list);
+		kfree(i7core_dev);
 	}
 	probed--;
 
